authentication/authentication.go

```diff
@@ -0,0 +1,153 @@
+	}
+
+	publicKeyResult, keyErr := loadEncryptionKey(encryptionKeyPath)
+	if keyErr != nil {
+		return "", &TokenError{Desc: "Error loading encryption key", From: keyErr}
+	}
+
+	opts := jose.SignerOptions{}
+	opts.WithType("JWT")
+	opts.WithHeader("kid", privateKeyResult.kid)
+
+	signer, err := jose.NewSigner(jose.SigningKey{Algorithm: jose.RS256, Key: privateKeyResult.key}, &opts)
+	if err != nil {
+		return "", &TokenError{Desc: "Error creating JWT signer", From: err}
+	}
+
+	encrypter, err := jose.NewEncrypter(
+		jose.A256GCM,
+		jose.Recipient{Algorithm: jose.RSA_OAEP, Key: publicKeyResult.key, KeyID: publicKeyResult.kid},
+		(&jose.EncrypterOptions{}).WithType("JWT").WithContentType("JWT"))
+
+	if err != nil {
+		return "", &TokenError{Desc: "Error creating JWT signer", From: err}
+	}
+
+	token, err := jwt.SignedAndEncrypted(signer, encrypter).Claims(rawClaims).CompactSerialize()
+
+	if err != nil {
+		return "", &TokenError{Desc: "Error signing and encrypting JWT", From: err}
+	}
+
+	return token, nil
+}
+
+// PublicKeyResult is a wrapper for the public key and the kid that identifies it
+type PublicKeyResult struct {
+	key *rsa.PublicKey
+	kid string
+}
+
+// PrivateKeyResult is a wrapper for the private key and the kid that identifies it
+type PrivateKeyResult struct {
+	key *rsa.PrivateKey
+	kid string
+}
+
+func loadSigningKey(signingKeyPath string) (*PrivateKeyResult, *KeyLoadError) {
+
+	keyData, err := ioutil.ReadFile(signingKeyPath)
+	if err != nil {
+		return nil, &KeyLoadError{Op: "read", Err: "Failed to read signing key from file: " + signingKeyPath}
+	}
+
+	block, _ := pem.Decode(keyData)
+	privateKey, err := x509.ParsePKCS1PrivateKey(block.Bytes)
+	if err != nil {
+		return nil, &KeyLoadError{Op: "parse", Err: "Failed to parse signing key from PEM"}
+	}
+
+	PublicKey, err := x509.MarshalPKIXPublicKey(&privateKey.PublicKey)
+	if err != nil {
+		return nil, &KeyLoadError{Op: "marshal", Err: "Failed to marshal public key"}
+	}
+
+	pubBytes := pem.EncodeToMemory(&pem.Block{
+		Type:  "PUBLIC KEY",
+		Bytes: PublicKey,
+	})
+
+	kid := fmt.Sprintf("%x", sha1.Sum(pubBytes))
+
+	return &PrivateKeyResult{privateKey, kid}, nil
+}
+
+func loadEncryptionKey(encryptionKeyPath string) (*PublicKeyResult, *KeyLoadError) {
+
+	keyData, err := ioutil.ReadFile(encryptionKeyPath)
+	if err != nil {
+		return nil, &KeyLoadError{Op: "read", Err: "Failed to read encryption key from file: " + encryptionKeyPath}
+	}
+
+	block, _ := pem.Decode(keyData)
+	pub, err := x509.ParsePKIXPublicKey(block.Bytes)
+	if err != nil {
+		return nil, &KeyLoadError{Op: "parse", Err: "Failed to parse encryption key PEM"}
+	}
+
+	kid := fmt.Sprintf("%x", sha1.Sum(keyData))
+
+	publicKey, ok := pub.(*rsa.PublicKey)
+	if !ok {
+		return nil, &KeyLoadError{Op: "cast", Err: "Failed to cast key to rsa.PublicKey"}
+	}
+
+	return &PublicKeyResult{publicKey, kid}, nil
+}
+
+// TokenError describes an error that can occur during JWT generation
+type TokenError struct {
```

Comment:
Why custom errors?

Reply:
The code started in go-launch a survey and I wanted to keep it as similar as possible .

Reply:
Ah, so it's Chris' fault ;)

Ideally you shouldn't use custom errors unless you're needing to capture specific extra data or use type assertions on the type of error returned from a function.